src/protocol/sort/reveal.rs

```diff
@@ -0,0 +1,107 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::{
+        mesh::{Gateway, Mesh},
+        prss::SpaceIndex,
+        Direction,
+    },
+    protocol::{context::ProtocolContext, RecordId, Step},
+    secret_sharing::Replicated,
+};
+use embed_doc_image::embed_doc_image;
+use serde::{Deserialize, Serialize};
+
+/// A message sent by each helper when they've revealed their own shares
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct AValue<F> {
+    a: F,
+}
+
+/// This implements reveal algorithm of "Efficient Secure Three-Party Sorting Protocol with an Honest Majority" at communication cost of 3R.
+/// Revealing is the protocol that on input a pair of shares, outputs a secret. For simplicity, we
+/// consider a simple revealing in which each Pi sends [a]i to Pi+1 and then reconstructs a from [a]i and
+/// [a]i−1.
+// Input: Each helpers know their own secret shares
+// Output: At the end of the protocol, all 3 helpers know a revealed (or opened) secret
+#[derive(Debug)]
+pub struct Reveal<F> {
+    input: Replicated<F>,
+}
+
+impl<F: Field> Reveal<F> {
+    #[allow(dead_code)]
+    pub fn new(input: Replicated<F>) -> Self {
+        Self { input }
+    }
+
+    /// Steps
+    /// Each helper sends their left share to the right helper. The helper then reconstructs their secret by adding the three shares
+    /// i.e. their own shares and received share.
+    #[embed_doc_image("reveal", "images/sort/reveal.png")]
+    #[allow(dead_code)]
+    pub async fn execute<M: Mesh, G: Gateway<M, S>, S: Step + SpaceIndex>(
+        self,
+        ctx: &ProtocolContext<'_, G, S>,
+        record_id: RecordId,
+        step: S,
+    ) -> Result<F, BoxError> {
+        let mut channel = ctx.gateway.get_channel(step);
+
+        let inputs = self.input.as_tuple();
+        // `to_helper.left` calculates part1 = (input.0 + input.1) - r1 and sends part1 to `to_helper.right`
+        channel
+            .send(
+                channel.identity().peer(Direction::Right),
+                record_id,
+                AValue { a: inputs.0 },
+            )
+            .await?;
+
+        // Sleep until `to_helper.left` sends us their part2 value
+        let AValue { a } = channel
+            .receive(channel.identity().peer(Direction::Left), record_id)
+            .await?;
+
+        Ok(inputs.0 + inputs.1 + a)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rand::rngs::mock::StepRng;
+    use tokio::try_join;
+
+    use crate::{
+        field::Fp31,
+        protocol::{sort::reveal::Reveal, QueryId, RecordId},
+        test_fixture::{make_contexts, make_world, share, TestStep, TestWorld},
+    };
+
+    #[tokio::test]
+    pub async fn reveal() {
+        let mut rand = StepRng::new(100, 1);
+
+        let input = Fp31::from(23_u128);
+        let share = share(input, &mut rand);
```

Comment:
As crazy as this sounds, what if there was a bug that only affected some numbers, but not 23. To avoid that we should generate a few random numbers in a loop and "reveal" all of them. Same for "reshare".